rmw_connext_cpp/src/functions.cpp

```diff
@@ -994,8 +994,17 @@ rmw_wait(rmw_subscriptions_t * subscriptions,
 
   // invoke wait until one of the conditions triggers
   DDSConditionSeq active_conditions;
-  DDS_Duration_t timeout = DDS_Duration_t::from_seconds(non_blocking ? 0 : 1);
+  DDS_Duration_t timeout;
+  bool non_blocking = false;
+  if (!wait_timeout) {
+    timeout = DDS_DURATION_INFINITE;
+  } else {
+    non_blocking = wait_timeout->sec == 0 && wait_timeout->nsec == 0;
+    timeout.sec = wait_timeout->sec;
+    timeout.nanosec = wait_timeout->nsec;
+  }
   DDS_ReturnCode_t status = DDS_RETCODE_TIMEOUT;
+
   while (DDS_RETCODE_TIMEOUT == status) {
```

Comment:
Double posting comment from https://github.com/ros2/rmw_opensplice/pull/55:

As I was writing a test for this, I realized that the existing logic of this loop is not what I thought it was.

Old behavior:

* Get "blocking" or "non-blocking in rclcpp.
* In rmw implementation:

* if non-blocking, call rmw_wait with a timeout of zero, and don't loop if timed out.
* if blocking: while return code is timeout, wait for 1 second. throw if error, continue if retcode was not timeout. The executor will always block indefinitely inside of rmw_wait.

Here's what I think the new behavior should be:

Get timeout from rclcpp.
* If NULL: block indefinitely in rmw_wait, by passing `DURATION_INFINITE` to `WaitSet::Wait`.
* If positive: pass the timeout value to `WaitSet::wait` and return if timed out.
* If non-blocking: same behavior as before.